File: src/easy/Q0028/Q0028.py
"""
28. Implement strStr()
https://leetcode.com/problems/implement-strstr/
"""


# Time: O(nm), Space: O(1), where n is the size of `haystack` and m is the size of `needle`.
def str_str(haystack: str, needle: str) -> int:
    if len(needle) == 0:
        return 0
    for i in range(len(haystack) - len(needle) + 1):
        if needle == haystack[i: i + len(needle)]:
            return i
    return -1

File: src/main/java/easy/Q0028.java
package easy;

import java.util.Objects;

/**
 * 28. Implement strStr()
 * https://leetcode.com/problems/implement-strstr/
 */
class Q0028 {

  // O(n^2) time, O(n) space, where n is the length of haystack
  int strStr(String haystack, String needle) {
    if (needle.isEmpty()) {
      return 0;
    }
    for (int i = 0; i < haystack.length() - needle.length() + 1; i++) {
      if (Objects.equals(haystack.substring(i, i + needle.length()), needle)) {  // .substring() is O(n) in time and space as of Java 7u6
        return i;
      }
    }
    return -1;
  }
}
